### DISCUSSION TITLE
Automated/scripted installs

### DISCUSSION THREAD

**Discussion by spusuf** [Jip-Hop/jailmaker#22] [Ideas]
The ability to run an install script that will automatically create a jail with the correct settings, prompt where needed, and run all the necessary installers for a specific application. 

That script can then easily be shared to help others easily install that application. Similar to the current state of TrueNAS core jails, where install scripts are the most popular and up-to-date way to install applications in jails.

Ideally once the functionality has been implemented a template will be provided on the github for others to follow, I am happy to make this template once the script is accommodating.

**Comment by Jip-Hop** [comment]
Thanks for the suggestions! I haven't used TrueNAS core jails myself but good to know you have, for comparisons sake. 😃

Which apps do you have in mind? I generally recommend/expect people will use docker inside the jail for this kind of convenience.

Once the jail has been created, whatever is done inside the jail is not really specific to jailmaker anymore. You can then use many generic tutorials or installation scripts (as you've noticed with the [install-docker.sh](https://get.docker.com/) convenience script).

**Comment by bcat** [comment]
Honestly, the big gap I see here is that the actual jail creation (`jlmkr create`) doesn't seem to be easily scriptable today.

Although I do intend to keep most of my apps in a single jail running Docker, there are some things I'd rather run in their own jails for more custom setup, better isolation, etc. And I like to document my system setup for my own personal use (e.g., so I have a record of what I did that I can easily redo, etc.). As such, I've been trying to make simple shell scripts that set up my own jails.

See [my Tailscale jail install script](https://gist.github.com/bcat/6409c7b0ef46b4e3a15e7d759c8e3f40) for an example. It's almost entirely automated, but the `jlmkr create` call in the middle prevents the flow from being completely smooth. It requires responding to several prompts from `jlmkr` and manually entering the correct values in cases where the defaults are not sufficient.

Honestly, if `jlmkr create` had a non-interactive mode that accepted all its input from either [command-line arguments](https://github.com/Jip-Hop/jailmaker/discussions/40) or a config file, that would super helpful for cases like this. :)

**Reply by Jip-Hop** [reply]
See also https://github.com/Jip-Hop/jailmaker/discussions/40

**Comment by Jip-Hop** [comment]
I'm locking this discussion as I've opened dedicated issues to track feature requests. Please see [Template for new jails](https://github.com/Jip-Hop/jailmaker/issues/84) and [Non-interactive CLI](https://github.com/Jip-Hop/jailmaker/issues/85).
